[DEV APPROVED] 9405 - Allow filter by Evidence type on Evidence Hub

scruti opened the pull request:
[TP-9405](https://moneyadviceservice.tpondemand.com/entity/9405-add-evidence-type-to-ev-hub)

- We want to be able to filter the Evidence Hub results by the "Evidence type".
- This PR also includes some refactor in the Search Form Parser in order to isolate responsibilities and allow future extension of the class.
- I recommend to review it Commit by Commit.

Page to check: (`/en/evidence_hub`)

## After the Changes

### Without selecting any option
![screenshot from 2018-09-06 10-36-14](https://user-images.githubusercontent.com/1227578/45149196-e6ded300-b1c0-11e8-9147-92a95add3c89.png)

### Filtering by a single type
![screenshot from 2018-09-06 10-36-47](https://user-images.githubusercontent.com/1227578/45149205-e8100000-b1c0-11e8-9f9b-b2ca02eb03c1.png)

### Filtering by multiple types
![screenshot from 2018-09-06 10-37-10](https://user-images.githubusercontent.com/1227578/45149207-ea725a00-b1c0-11e8-8a9f-0673abd3050e.png)

tomas-stefano commented on features/evidence_hub_search.feature:
```diff
@@ -84,6 +84,18 @@ Feature: Evidence Hub Search
+      | countries           | United Kingdom              |
+      | year of publication | 2017                        |
+    And I should see the "first" evidence summary icon linking to "Evaluation" article
+
```
I think is worth to add a search by insight, review or two altogether. Wdyt?

scruti replied:
Sure! Going to add it.

your commented on app/services/year_of_publication_parsing.rb:
```diff
@@ -0,0 +1,29 @@
+module YearOfPublicationParsing
```
Love this extraction!

Only thing, i would argue that maybe `YearOfPublicationParser` would fit better.

Plus, if this parser will be used only by the `SearchFormParser`, maybe you could namespace these two files so that it's clear they go together (and the `services` folder does not become messy).

scruti replied:
Oh got a little bit confused with the naming as originally was adding everything under the module for namespacing (without intending a service object), then ended being a service object and didn't realize the name was out of convention. Good catch @your  and @MarcoBgn.

+1 for the namespacing too. Originally I thought about namespacing both under `evidence_hub`, may help to a possible Domain Design gouping in the future (as `evidence_hub`  seems to be a domain concept around different bits of the app).

MarcoBgn commented:
Nice work @scruti ! 

I agree with Giuseppe and would name it `YearOfPublicationParser`, what do you think?

scruti commented:
@tomas-stefano @MarcoBgn @your, I just rebased with these changes:
- Added test covering for multiselection in the search filter.
- fixed the wrong naming in the Service Object.
- namespaced the both parser service objects under `evidence_hub`.

Thanks for all the suggestions, let me know if you would change anything else!

masbuild commented:
<!--
  0 Errors
  1 Warning: This PR is huge. Consider next...
  0 Messages
  1 Markdown
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is huge. Consider next time to split up the workinto multiple PRs. A good guide about small PRs can be seenhere(link to mas-standards - small PRs guide)</td>
    </tr>
  </tbody>
</table>

At Money Advice Service, we use Danger gem, to ensure that our commits and pull requests follow [standards](https://github.com/moneyadviceservice/mas-standards/blob/master/Dangerfile). If you think one of the points here has been raised in error (or shouldn't be in our standards), open an issue or pull request on [our Danger config](https://github.com/moneyadviceservice/mas-standards).

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

scruti commented:
About Danger "huge PR" message I intend to ignore it if is right with the team standards as is just:
- refactor for better extensibility.
- extend the code adding the required feature.

Don't think is a massive thing to review and both commits make sense together. 

Feel free to raise you concerns if you don't agree with this decision :P
